### PR TITLE
feat(data): add Iris dataset provider (#1044)

### LINKED ISSUE #1044: [Feature]: Add an Iris dataset provider (and a tabular RawDataset -> Dataset bridge)
> 👋 **New to SKaiNET? This one is for you.**
>
> This issue is deliberately written to be a *first* contribution. It touches one module, needs
> no maths or GPU knowledge, and there is an existing file in the codebase you can copy the
> structure from almost line by line. If anything below is unclear, that is a bug in the issue,
> not in you — please just ask in the comments and we will fix the wording.
>
> **Estimated effort:** half a day for Part 1. **Prerequisites:** Kotlin basics. That is it.

---

# 🧠 D: DOCUMENT — Problem & Opportunity

**Summary (1–2 sentences):**
> SKaiNET ships ready-made dataset providers for MNIST, Fashion-MNIST and CIFAR-10, but nothing
> for small tabular datasets. Anyone writing a classification example has to hand-write CSV
> parsing, one-hot encoding and tensor batching before they can touch the interesting part.

## What is missing

Loading MNIST in SKaiNET today is one line:

```kotlin
val train = MNIST.loadTrain()   // suspend; downloads, caches, returns a Dataset
```

There is no equivalent for Iris — or for *any* tabular dataset. The pieces exist but do not meet:

- ✅ `skainet-data-source` parses CSV into a `RawDataset` (a `List<RawDataRow>`, where each row is
  a `Map<String, String>`, plus a `DataSchema`).
- ✅ `skainet-data-api` defines `Dataset<X, Y>`, which gives you `split`, `shuffle`, `batches`,
  `epochs`, `filter` and `mapX`/`mapY` for free.
- ❌ **Nothing converts the first into the second.** `RawDataset` holds strings; `Dataset` must
  emit `DataBatch` tensors. That conversion is left to every user.

## Why this matters now

PR [#34](https://github.com/SKaiNET-developers/SKaiNET-examples/pull/34) from [@EdinEd](https://github.com/EdinEd) added a lovely Iris
multiclass example — and roughly half of its ~200 lines are data plumbing:

| Hand-written in the example | Already in SKaiNET |
|---|---|
| `loadKaggleCsv()` | `DataFormat.CSV` parser in `skainet-data-source` |
| `test_train_split()` | `Dataset.split(ratio, seed, stratified)` |
| `labelToOneHot()` | — (genuinely missing) |
| manual per-sample tensor construction | `Dataset.batches()` |

The same boilerplate is being reproduced in the accompanying teaching notebook, where it is
actively in the way — the notebook has to spend a whole section on `Dataset` plumbing before it
can talk about classification.

## Who benefits

- **Newcomers** — the first Kotlin ML example someone reads shouldn't open with a CSV parser.
- **Example and notebook authors** — one line instead of forty, in every tabular example.
- **Everyone with a CSV** — Part 2 below makes *any* tabular file loadable, not just Iris.

---

# 🔍 A: ASSESS — Feasibility & Impact

### ✔️ Feasibility

**High.** This is additive: no existing API changes, nothing to deprecate, no numerical work.
`MNISTDataset` in
`skainet-data/skainet-data-simple/src/commonMain/kotlin/sk/ainet/data/mnist/MNISTData.kt`
is a direct template — Iris is strictly simpler (no gzip, no IDX binary format, no image layout).

Iris is ~4 KB and public domain, so unlike MNIST it can simply be **embedded as a resource**.
No download, no cache directory, no network in tests. That removes most of what makes
`MNISTLoader` complicated, which is exactly why this makes a good first issue.

### ✔️ Expected Impact

- Every tabular example and notebook loses its boilerplate section.
- Part 2 unlocks CSV/TSV/JSON-Lines → `Dataset` generally, which is a frequently requested path.
- Gives SKaiNET the "toy dataset" on-ramp that scikit-learn's `load_iris()` provides.

### ✔️ Risks / Constraints

| Risk | Mitigation |
|---|---|
| Class-index ordering must be stable | Fix alphabetical order in a documented constant; assert in tests |
| `Dataset` `Y` type choice affects stratification | Use `Int` class index — see the trap in Research below |
| Feature column order must be stable | Define explicitly, never rely on `Map` iteration order |
| Embedding data in a jar | 4 KB, public domain — negligible |

### ✔️ Dependencies

- **Part 1:** `skainet-data-api` (for `Dataset`/`DataBatch`) and `skainet-lang-core` (tensors).
  Lands in `skainet-data-simple` beside the MNIST/CIFAR providers.
- **Part 2:** additionally `skainet-data-source` (for `RawDataset`).
- No third-party libraries.

---

# 📚 R: RESEARCH — What Must Be Understood First?

### Reading list (all in this repo — start here)

- [ ] `skainet-data/skainet-data-api/src/commonMain/kotlin/sk/ainet/data/Dataset.kt` — **the one
      file to actually read**. Note which members are `abstract` (you must implement them) versus
      `open`/concrete (you get them free).
- [ ] `skainet-data/skainet-data-api/src/commonMain/kotlin/sk/ainet/data/DataBatch.kt` — what you
      must produce.
- [ ] `skainet-data/skainet-data-simple/.../mnist/MNISTData.kt` — the pattern to copy.
- [ ] `skainet-data/skainet-data-simple/.../mnist/MNIST.kt` — the tiny public entry-point object.
- [ ] `skainet-data/skainet-data-source/.../DataFormatParser.kt` — `RawDataset` shape (Part 2 only).

### ⚠️ Three traps worth knowing before you start

**1. You must override `createIndexedDataBatch`, not just `createDataBatch`.**
`split()` and `shuffle()` return an *index view* over your dataset rather than copying data. When
that view asks for a batch, it passes **non-contiguous** indices. The base-class default only
handles contiguous ranges and throws otherwise. Override both and shuffled/split/filtered views
all work; override only the first and `split(...)` then `batches(...)` fails at runtime.

**2. `Y` should be `Int` (the class index), not the one-hot `FloatArray`.**
Stratified splitting buckets samples in a `LinkedHashMap` keyed by `getY(idx)`. `FloatArray`
compares by *identity*, so a one-hot label would put every sample in its own bucket and silently
defeat stratification — no error, just a split that isn't stratified. Keep `getY` returning the
class index; do the one-hot conversion inside batch construction.

**3. Cross-entropy expects logits, not probabilities.**
Not your code, but relevant to the tests: `CategoricalCrossEntropyLoss` applies log-softmax
internally. If you write a smoke test that trains a small model, do **not** put a softmax on the
output layer.

### Prior art

- [ ] scikit-learn `load_iris()` — returns data/target/feature_names/target_names. Worth mirroring
      the *metadata* idea: expose `featureNames` and `classNames`, not just numbers.
- [ ] PyTorch `torchvision.datasets` — `Dataset` + transform separation, which SKaiNET already follows.

### Open questions (comment if you have an opinion)

> - Should `Iris.load()` be `suspend`? MNIST's is, because it downloads. Embedded Iris needn't be —
>   but a non-suspend outlier breaks call-site symmetry. **Current lean:** make it `suspend` for
>   consistency, since callers are usually in `runBlocking` anyway.
> - Should the provider standardise features, or return raw centimetres? **Current lean:** raw.
>   Scaling statistics must be fitted on the training split only, so the provider is the wrong
>   place — it doesn't know where you'll split.
> - Does Part 2's adapter belong in `skainet-data-source` or a new `skainet-data-tabular` module?

---

# 🛠️ C: CODE — Implementation Plan

**Part 1 and Part 2 are independent.** Part 1 alone is a complete, mergeable contribution — please
do not feel obliged to take both. If you only want Part 1, say so in a comment and someone else
can pick up Part 2.

## Part 1 — `Iris` provider *(the good-first-issue part)*

**Location:** `skainet-data/skainet-data-simple/src/commonMain/kotlin/sk/ainet/data/iris/`

### Proposed API

```kotlin
package sk.ainet.data.iris

/** One Iris flower: four measurements in cm plus its species index. */
public data class IrisSample(
    val sepalLength: Float,
    val sepalWidth: Float,
    val petalLength: Float,
    val petalWidth: Float,
    val label: Int          // 0=setosa, 1=versicolor, 2=virginica
)

/**
 * Iris dataset (Fisher, 1936). 150 samples, 4 features, 3 balanced classes.
 *
 *   x -> Tensor<FP32, Float> [batch, 4]
 *   y -> Tensor<FP32, Float> [batch, 3]  (one-hot)
 */
public data class IrisDataset(
    val samples: List<IrisSample>,
    private val executionContext: ExecutionContext = DefaultDataExecutionContext()
) : Dataset<FloatArray, Int>() { /* ... */ }

/** Entry point, mirroring `MNIST`. */
public object Iris {
    public val featureNames: List<String> =
        listOf("sepalLength", "sepalWidth", "petalLength", "petalWidth")

    public val classNames: List<String> =
        listOf("Iris-setosa", "Iris-versicolor", "Iris-virginica")

    /** Loads the bundled copy of the dataset. */
    public suspend fun load(): IrisDataset
}
```

Usage — replacing ~40 lines with one:

```kotlin
val (train, test) = Iris.load().split(0.8, seed = 42L, stratified = true)
```

### Development tasks

- [ ] Add `iris.csv` (150 rows, public domain) as a resource in `skainet-data-simple`
- [ ] `IrisSample` data class
- [ ] `IrisDataset : Dataset<FloatArray, Int>` implementing `xSize`, `getX`, `getY`, `shuffle()`,
      `split(ratio)`, `createDataBatch(...)` **and `createIndexedDataBatch(...)`** (trap 1 above)
- [ ] `Iris` object with `load()`, `featureNames`, `classNames`
- [ ] Register the module in the docs' module table if needed

### Tests

- [ ] `load()` returns exactly 150 samples, 50 per class
- [ ] Feature values fall in the documented ranges (sepal length 4.3–7.9, petal width 0.1–2.5)
- [ ] Class indices agree with `classNames` ordering
- [ ] `dataBatch` produces tensors of shape `[n, 4]` and `[n, 3]`; each `y` row sums to 1.0
- [ ] `split(0.8, seed = 42L, stratified = true)` gives 120/30 with 40/10 per class
- [ ] **Regression test for trap 1:** `split(...)` *then* `batches(...)` must not throw
- [ ] Same seed ⇒ identical splits across runs

### Acceptance criteria

- [ ] `Iris.load()` works with no network access
- [ ] Public API has KDoc; follows the `MNIST` naming pattern
- [ ] Passes `explicit-api` and binary-compatibility-validator checks
- [ ] No regressions elsewhere

## Part 2 — tabular `RawDataset` → `Dataset` bridge *(follow-up; separate PR welcome)*

Generalises Part 1 so any CSV/TSV/JSON-Lines source becomes a `Dataset`.

```kotlin
val dataset: Dataset<FloatArray, Int> = rawDataset { from("file:///data/wine.csv") }
    .toTabularDataset {
        features("alcohol", "malicAcid", "ash")   // column order is explicit
        label("cultivar")                          // string labels -> stable indices
        oneHotLabels()                             // or labelIndices() for Int32 targets
    }
```

- [ ] Design the builder scope (`features`, `label`, `oneHotLabels`/`labelIndices`)
- [ ] Implement `TabularDataset : Dataset<FloatArray, Int>` (Part 1's class, generalised)
- [ ] Decide module placement (see Open Questions)
- [ ] Handle missing values and unparseable cells with a clear error naming row *and* column
- [ ] Tests: numeric + string labels, missing columns, malformed rows, schema mismatch
- [ ] Reimplement Part 1's `IrisDataset` on top of it, keeping the `Iris.load()` API unchanged
- [ ] Add a `how-to` page under `docs/modules/ROOT/pages/how-to/`

---

# 💬 Additional Notes

### Reference implementation

A complete working `IrisDataset` — including both `createDataBatch` overrides — already exists in
the multiclass-classification teaching notebook that accompanies this issue. It is written against
released **0.40.1** and can be lifted almost verbatim into `skainet-data-simple`. Treat it as a
starting point, not a specification.

### Why this is a good first issue

- **One module, additive.** Nothing existing changes, so review is low-risk.
- **A template to copy.** `MNISTData.kt` does the same job in a harder setting.
- **Obvious tests.** Iris has 150 rows and known-good values; assertions write themselves.
- **Visible payoff.** Your PR deletes forty lines from every tabular example that follows.

### Getting started

```bash
git clone https://github.com/SKaiNET-developers/SKaiNET.git
cd SKaiNET
./gradlew :skainet-data:skainet-data-simple:jvmTest
```

See [`CONTRIBUTING.md`](../CONTRIBUTING.md) and [`GITFLOW.adoc`](../GITFLOW.adoc) — branch off
`develop`. Please comment before starting so we can assign it to you and avoid duplicate work.

### Related

- SKaiNET-examples PR #34 — the Iris example that motivated this
- Companion notebook: *Multiclass Classification with SKaiNET*
- Existing providers: `sk.ainet.data.mnist`, `sk.ainet.data.fashionmnist`, `sk.ainet.data.cifar10`

### PR DESCRIPTION
**Pull request by AjithGoveas**
## What

Adds a first-class Iris dataset provider to `skainet-data-simple`, enabling one-line access to the 1936 Fisher Iris dataset for classification tutorials.

```kotlin
val (train, test) = Iris.load().split(0.8, seed = 42L, stratified = true)
```

## Files changed
- Iris.kt — entry object: featureNames, classNames, suspend fun load()
- IrisData.kt — IrisSample data class, embedded 150-row CSV (UCI canonical ordering, public domain), strict parser with error messages naming line + column
- IrisDataset.kt — Dataset<FloatArray, Int>: FP32 [n,4] features, FP32 [n,3] one-hot labels. Both createDataBatch and createIndexedDataBatch overridden so split()/shuffle() views batch correctly
- IrisDatasetTest.kt — 8 tests in jvmTest (same convention as MNIST/CIFAR tests)
- README.md — Iris listed in built-in loaders
- data-sources-getting-started.adoc — Iris usage snippet added
## Design notes
- Data is source-embedded (no file resource) to stay platform-agnostic including JS and Wasm browser targets, matching the repo's existing approach
- load() is suspend for call-site symmetry with MNIST.loadTrain() / CIFAR10.load()
- Raw centimetres returned; no standardisation (scaling should be fitted on training split only)
- getY() returns Int, not one-hot — one-hot conversion is deferred to batch construction so stratified splitting works correctly
## Checklist
- Closes #1044
- All 8 Iris tests pass
- No regressions in existing skainet-data-simple tests
- Passes explicit-api compilation
- Public API has KDoc

### CODE REVIEW

**Review by michalharakal — COMMENTED** [review]
First of all: welcome, and thank you, the quality bar you set here is genuinely impressive. 
The `createIndexedDataBatch` override for non-contiguous views, deferring one-hot to batch construction so stratified splitting works, and the test proving the split-then-batch path — that's exactly the subtle stuff that's usually missed. 🎉

I pushed one tiny fix directly to your branch (0dd46208) so we can merge without another round-trip: in the unknown-species error message, `$Iris.classNames` interpolated `Iris.toString()` plus the literal text `.classNames` — it needed braces, `${Iris.classNames}`. I also added an assertion to the parser test so it would catch this in future.

Two purely optional, non-blocking notes — fine as follow-ups or to just leave as-is:

1. The no-arg `shuffle()` override uses `Random.Default`, so it's non-deterministic. Probably intentional, but a short KDoc note pointing reproducibility-minded users at the seeded `shuffle(seed)` overload would help.
2. The plain `split(splitRatio)` cuts at a raw index, and since the embedded data is ordered by species (50 per class), an unstratified 0.8 split puts every virginica into the test set. A KDoc warning recommending `split(ratio, seed, stratified = true)` would save someone a confusing afternoon.

Thanks again — looking forward to more! I am preparing another similar `first good issues` so we can reach funstion/Ops parity with KotlinDL :-)  #290

**Review by michalharakal — APPROVED** [review]
